frontend/public/components/utils/kebab.tsx

```diff
@@ -207,7 +207,9 @@ export type KebabOption = {
   href?: string, callback?: () => any;
   accessReview?: AccessReviewResourceAttributes;
 };
-export type KebabAction = (kind: K8sKind, obj: K8sResourceKind) => KebabOption;
+export type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption;
+
+export type KebabOptionsCreator = (kindObj: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption[];
```

Comment:
@suomiy @rawagner @spadgett 

`KebabAction` is actually the "creator" of `KebabOption` (single) object, while the proposed `KebabOptionsCreator` is the "creator" of `KebabOption` object array, reusing the same argument list.

This adds ambiguity as to which one to use, since it's either a list of functions (`KebabAction[]`) or a single function (`KebabOptionsCreator`), each of which have different return types (and therefore different signatures), and therefore need to be handled differently.

We can make this simpler, without introducing an alternative type/signature

```ts
export type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption | KebabOption[];
```

and in `PageHeading` component

```tsx
const hasMenuActions = !_.isEmpty(menuActions); // no change here, contract stays the same
// { hasMenuActions && <ActionsMenu actions={menuActions.map(a => a(kindObj, data))} /> }
{ hasMenuActions && <ActionsMenu actions={menuActions.reduce((items: KebabOption[], action: KebabAction) => [
  ...items,
  ...(_.castArray(action(kindObj, data))),
], [])} /> }
```

This way, we use a single `KebabAction` type while supporting multiple outputs, i.e. `KebabOption | KebabOption[]`.

Reply:
Yup, that would be nicer but it would not work.

The issue with this approach

- the resources are not the actual resources but are computed from the loaded data, please see for example vm status https://github.com/openshift/console/pull/1906/files#diff-e1ea60ed2e820073acdfd59753ed6636R147
- the resources are first accessible in PageHeading (please see DetailsPage)
- we have to compute something from these resources, hence the need for callback

Reply:
Both `KebabAction` and `KebabOptionsCreator` are functions with same arguments

```
kind: K8sKind, obj: K8sResourceKind, resources?: any
```

but different return value types, `KebabOption` vs. `KebabOption[]`.

The argument semantics when calling one function vs. the other one should be the same, right?

```js
_.isFunction(menuActions)
  ? menuActions(kindObj, data, extraResources) // case A
  : menuActions.map(a => a(kindObj, data, extraResources)) // case B
```

- In case A, we run one function and expect `KebabOption[]`.
- In case B, we run N functions each returning `KebabOption` and map result to `KebabOption[]`.
- The discriminator is the `menuActions` shape, either a single function (case A) or array of functions (case B).

---

**TL;DR**

In both `PageHeadingProps` and `DetailsPageProps`:

```diff
-menuActions?: any[];
+menuActions?: Function[] | KebabMultiAction; // FIXME: should be "KebabAction[] | KebabMultiAction"
```

At the bottom of `headings.tsx` (this type is unrelated to `kebab.tsx`, we need it specifically for `PageHeading`):

```ts
export type KebabMultiAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption[];
```